Objects/weakrefobject.c

```diff
@@ -958,8 +983,11 @@ PyObject_ClearWeakRefs(PyObject *object)
     /* Remove the callback-less basic and proxy references */
     if (*list != NULL && (*list)->wr_callback == NULL) {
         clear_weakref(*list);
-        if (*list != NULL && (*list)->wr_callback == NULL)
+        remove_weakref_from_referent(*list);
```

Comment:
Here we already have the list, so we should use `weaklist_remove_weakref()` directly:

```suggestion
        weaklist_remove_weakref(list, *list);
```